File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.py
class Solution:
    def dp(self, prices, currentDay, remainingTransactions, holding, lookup):
        # Return 0 if there aren't any transactions left or there are no more stocks.
        if remainingTransactions == 0 or currentDay == len(prices):
            return 0
        if not lookup[currentDay][remainingTransactions][holding]:
            if holding:
                # Return max(did nothing on current day, sold all the stocks on current day).
                lookup[currentDay][remainingTransactions][holding] = max(self.dp(prices, currentDay+1, remainingTransactions, holding, lookup), prices[currentDay] + self.dp(prices, currentDay+1, remainingTransactions-1, 0, lookup))
            else:
                # Return max(did nothing on current day, bought all the stocks on current day).
                lookup[currentDay][remainingTransactions][holding] = max(self.dp(prices, currentDay+1, remainingTransactions, holding, lookup), -prices[currentDay] + self.dp(prices, currentDay+1, remainingTransactions, 1, lookup))
        return lookup[currentDay][remainingTransactions][holding]
            
    def maxProfit(self, k: int, prices: List[int]) -> int:
        return self.dp(prices, 0, k, 0, [[[None for _ in range(2)] for _ in range(k+1)] for _ in range(len(prices))])

File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.cpp
class Solution {
  public:
    static constexpr int MAXN = 1e3 + 5;
    static constexpr int MAXM = 1e2 + 5;
    int cache[MAXN][MAXM][2];
    int n, m;
    vector <int> prices;
    int solve(int i, int j, bool bought) {
        if(i == n)
            return 0;
        int &ret = cache[i][j][bought];
        if(~ret)return ret;
        ret = solve(i + 1, j, bought);
        if(bought) {
            ret = max(ret, solve(i + 1, j, false) + prices[i]);
        } else {
            if(j > 0)
             ret = max(ret, solve(i + 1, j - 1, true) - prices[i]);
        }
        return ret;
    }

    int maxProfit(int k, vector<int>& _prices) {
        prices = _prices;
        n = int(prices.size());
        m = k;
        memset(cache, -1, sizeof cache);
        return solve(0, m, false);
    }
};